proof/proof.go

```diff
@@ -0,0 +1,160 @@
+// Copyright 2022 Google LLC. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package proof contains helpers for constructing log Merkle tree proofs.
+package proof
+
+import (
+	"errors"
+	"fmt"
+	"math/bits"
+
+	"github.com/transparency-dev/merkle/compact"
+)
+
+// Nodes contains information on how to construct a log Merkle tree proof.
+type Nodes struct {
+	// IDs contains the IDs of non-ephemeral nodes sufficient to build the proof.
+	IDs []compact.NodeID
+	// begin is the beginning index (inclusive) into the IDs[begin:end] subslice
+	// of the nodes comprising the ephemeral node.
+	begin int
+	// end is the ending (exclusive) index into the IDs[begin:end] subslice of
+	// the nodes comprising the ephemeral node.
+	end int
+}
+
+// Inclusion returns the information on how to fetch and construct an inclusion
+// proof for the given leaf index in a log Merkle tree of the given size.
+func Inclusion(index, size uint64) (Nodes, error) {
+	if index >= size {
+		return Nodes{}, fmt.Errorf("index %d out of bounds for tree size %d", index, size)
+	}
+	return nodes(index, 0, size), nil
+}
+
+// Consistency returns the information on how to fetch and construct a
+// consistency proof between the two given tree sizes of a log Merkle tree.
+func Consistency(size1, size2 uint64) (Nodes, error) {
+	if size1 > size2 {
+		return Nodes{}, fmt.Errorf("tree size %d > %d", size1, size2)
+	}
+	if size1 == size2 || size1 == 0 {
+		return Nodes{IDs: []compact.NodeID{}}, nil
+	}
+
+	// TODO(pavelkalinnikov): Make the capacity estimate accurate.
+	proof := make([]compact.NodeID, 0, bits.Len64(size2)+1)
+
+	// Find the biggest perfect subtree that ends at size1.
+	level := uint(bits.TrailingZeros64(size1))
+	index := (size1 - 1) >> level
+	// If it does not cover the whole size1 tree, add this node to the proof.
+	if index != 0 {
+		proof = append(proof, compact.NewNodeID(level, index))
+	}
+
+	// Now append the path from this node to the root of size2.
+	p := nodes(index, level, size2)
+	p.IDs = append(proof, p.IDs...)
+	if len(proof) == 1 && p.begin < p.end {
```

Comment:
Maybe just add a comment here to explain that this is adjusting for when we already have a node in the proof for the cases where size1 isn't a power of two (and the verifier needs to be able to re-create the root hash at size1)?